docs/source/format/CDeviceDataInterface.rst

```diff
@@ -650,6 +652,137 @@ The stream source is not assumed to be thread-safe. Consumers wanting to
+    If successfully handled, the function should return 0, otherwise it should return an
+    ``errno``-compatible error code.
+
+    The ``void*`` parameter exists for producers to provide any extra contextual information
+    they want.
+
+    Unless the ``on_error`` handler is called, this will always get called at least once. The
+    end of the stream is indicated by it being called with an empty or released ``ArrowDeviceArray``.
+
+.. c:member:: void (*ArrowAsyncDeviceStreamHandler.on_error)(struct ArrowAsyncDeviceStreamHandler, int, const char*, const char*)
+
+    *Mandatory.* Handler to be called when an error is encountered by the producer. After calling
+    this, the ``release`` callback will be called as the last call on this struct. The parameters
+    are an ``errno``-compatible error code and an optional error message and metadata.
+
+    If the message and metadata are not ``NULL``, their lifetime is only valid during the scope
+    of this call. A consumer who wants to maintain these values past the return of this function
+    *MUST* copy the values themselves.
+
+    If the metadata parameter is not ``NULL``, to provide key-value error metadata, then it should
+    be encoded identically to the way that metadata is encoded in :c:member:`ArrowSchema.metadata`.
+
+.. c:member:: void (*ArrowAsyncDeviceStreamHandler.release)(struct ArrowAsyncDeviceStreamHandler*)
+
+    *Mandatory.* A pointer to a consumer-provided release callback for the handler.
+
+.. c:member:: void* ArrowAsyncDeviceStreamHandler.private_data
+
+    *Optional.* An opaque pointer to consumer-provided private data.
+
+    Producers *MUST NOT* process this member. Lifetime of this member is handled by
+    the consumer, and especially by the release callback.
+
+Result lifetimes
+''''''''''''''''
+
+The ``ArrowSchema`` and ``ArrowDeviceArray`` that are passed to the ``on_schema`` and
+``on_next`` callbacks must be released independently. Their lifetimes are not tied to
+the ``ArrowAsyncDeviceStreamHandler`` or the handler functions.
+
+The ``const char*`` error ``message`` and ``metadata`` which are passed to ``on_error``
+are only valid within the scope of the ``on_error`` function itself. They must be copied
+if it is necessary for them to exist after it returns.
+
+Stream Handler Lifetime
+'''''''''''''''''''''''
+
+Lifetime of the async stream handler is managed using a release callback with similar
+usage as in :ref:`C data interface <c-data-interface-released>`.
+
+Thread safety
+'''''''''''''
+
+All handler functions should only be called in a serialized manner, but are not guaranteed
+to be called from the same thread every time. A producer should wait for handler callbacks to
+return before calling the next handler callback. As a result, a consumer can manage back-pressure
+simply by sleeping or otherwise waiting to return from the ``on_next`` handler if necessary.
```

Comment:
Do we really want to encourage consumers to handle backpressure by sleeping? This forces the producer to invoke consumer callbacks from a dedicated thread, adding complexity and overhead.